### ISSUE TITLE
Reproducible segfault in OSX when maximizing a buffer that has a terminal above it

### ISSUE THREAD

**Issue by ghamerly** [vim/vim#6332]
I run vim (currently version 8.2.1000_1) on OSX (10.14.6), installed via Homebrew.

For the last month or more (not sure), over several updated builds of vim, I can reliably cause a segfault in vim by doing the following:

1. open an empty file (`vim --clean`)
2. run `:term` to open a terminal window (above the buffer)
3. switch the cursor back to the empty file buffer (`ctrl-w ctrl-w`)
4. maximize the empty file buffer (`ctrl-w _`)

Then vim crashes with
```
Vim: Caught deadly signal SEGV
Vim: Finished.
Segmentation fault: 11
```

The same thing happens if there are other buffers open. It doesn't seem to matter if the buffers are empty or have contents. The one constant seems to be "maximize a buffer which has a terminal window immediately above it".

I ran vim under `lldb` and got the following stack trace when it crashes:

```
(lldb) bt                                                                       
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x166636174)
  * frame #0: 0x000000010011d5db vim`char_needs_redraw + 49                     
    frame #1: 0x000000010011cb37 vim`screen_line + 465                          
    frame #2: 0x00000001001663b1 vim`term_update_window + 1150                  
    frame #3: 0x0000000100031ae1 vim`win_update + 228                           
    frame #4: 0x0000000100031882 vim`update_screen + 1032                       
    frame #5: 0x000000010003493d vim`redraw_after_callback + 59                 
    frame #6: 0x0000000100163420 vim`write_to_term + 389                        
    frame #7: 0x00000001001cacc1 vim`may_invoke_callback + 1369                 
    frame #8: 0x00000001001cc1d9 vim`channel_parse_messages + 351               
    frame #9: 0x0000000100089459 vim`parse_queued_messages + 148                
    frame #10: 0x00000001001769d5 vim`inchar_loop + 96
    frame #11: 0x0000000100176931 vim`ui_inchar + 161
    frame #12: 0x0000000100085b1f vim`inchar + 489
    frame #13: 0x0000000100087e4f vim`vgetorpeek + 1054
    frame #14: 0x00000001000875b3 vim`vgetc + 309
    frame #15: 0x0000000100089005 vim`safe_vgetc + 9
    frame #16: 0x00000001000c5140 vim`normal_cmd + 292
    frame #17: 0x00000001001dcb30 vim`main_loop + 1516
    frame #18: 0x00000001001dc241 vim`vim_main2 + 3285
    frame #19: 0x00000001001db1de vim`main + 6250
    frame #20: 0x00007fff66df23d5 libdyld.dylib`start + 1
(lldb)  
```

**Expected behavior**

Not a crash. The same actions do not crash under any other environments I use (these other environments are all linux).

**Environment (please complete the following information):**
Vim version:
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jun 22 2020 22:29:15)
macOS version
Included patches: 1-1000
Compiled by Homebrew
Huge version without GUI.  Features included (+) or not (-):
+acl               -farsi             +mouse_sgr         +tag_binary
+arabic            +file_in_path      -mouse_sysmouse    -tag_old_static
+autocmd           +find_in_path      +mouse_urxvt       -tag_any_white
+autochdir         +float             +mouse_xterm       -tcl
-autoservername    +folding           +multi_byte        +termguicolors
-balloon_eval      -footer            +multi_lang        +terminal
+balloon_eval_term +fork()            -mzscheme          +terminfo
-browse            +gettext           +netbeans_intg     +termresponse
++builtin_terms    -hangul_input      +num64             +textobjects
+byte_offset       +iconv             +packages          +textprop
+channel           +insert_expand     +path_extra        +timers
+cindent           +ipv6              +perl              +title
-clientserver      +job               +persistent_undo   -toolbar
+clipboard         +jumplist          +popupwin          +user_commands
+cmdline_compl     +keymap            +postscript        +vartabs
+cmdline_hist      +lambda            +printer           +vertsplit
+cmdline_info      +langmap           +profile           +virtualedit
+comments          +libcall           -python            +visual
+conceal           +linebreak         +python3           +visualextra
+cryptv            +lispindent        +quickfix          +viminfo
+cscope            +listcmds          +reltime           +vreplace
+cursorbind        +localmap          +rightleft         +wildignore
+cursorshape       +lua               +ruby              +wildmenu
+dialog_con        +menu              +scrollbind        +windows
+diff              +mksession         +signs             +writebackup
+digraphs          +modify_fname      +smartindent       -X11
-dnd               +mouse             -sound             -xfontset
-ebcdic            -mouseshape        +spell             -xim
+emacs_tags        +mouse_dec         +startuptime       -xpm
+eval              -mouse_gpm         +statusline        -xsmp
+ex_extra          -mouse_jsbterm     -sun_workshop      -xterm_clipboard
+extra_search      +mouse_netterm     +syntax            -xterm_save
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: clang -c -I. -Iproto -DHAVE_CONFIG_H   -DMACOS_X -DMACOS_X_DARWIN  -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: clang   -L. -fstack-protector-strong -L/usr/local/lib -L/usr/local/opt/libyaml/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/readline/lib  -L/usr/local/lib -o vim        -lncurses -liconv -lintl -framework AppKit  -L/usr/local/opt/lua/lib -llua5.3 -mmacosx-version-min=10.14 -fstack-protector-strong -L/usr/local/lib  -L/usr/local/Cellar/perl/5.32.0/lib/perl5/5.32.0/darwin-thread-multi-2level/CORE -lperl -lm -lutil -lc  -L/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/config-3.8-darwin -lpython3.8 -framework CoreFoundation  -lruby.2.7     
```
 - OS: MacOS 10.14.6
 - Terminal: both xterm (under XQuartz) and MacOS Terminal.app

** Other information **

On the version that crashes, similar actions do *not* cause a crash if the windows are arranged in any other configuration (side-by-side (terminal on either side), or with the buffer *above* the terminal window).

I have not successfully built vim on my own machine, so I have been unable to do any search for when this bug began.

**Comment by dpelle** [comment]
This is probably the same issue as the one fixed in patch 8.2.1030:
```
commit da58134eedf43ae4b9013c93ecbdf55e4da4b8a3
Author: Bram Moolenaar <Bram@vim.org>
Date:   Sun Jun 21 17:57:32 2020 +0200

    patch 8.2.1030: reducing size of a terminal window may cause a crash
    
    Problem:    Reducing size of a terminal window may cause a crash.
    Solution:   Make sure the row and column don't become negative. (closes #6273)
```
Can you try the latest version of Vim with that fix?

**Comment by ghamerly** [comment]
Thanks!

I was able to build and run da58134eedf43ae4b9013c93ecbdf55e4da4b8a3, and the error is gone.

I also built the commit before that (23c55273739b2fb09912140e2e03a4b8ff15dd04) and the error exists.

So indeed this was fixed by da58134eedf43ae4b9013c93ecbdf55e4da4b8a3. Thanks for your help @dpelle.
